### ISSUE TITLE
Support for Option arguments of user types

### ISSUE THREAD

**Issue by fjarri** [rustwasm/wasm-bindgen#2370]
Currently, having a binding that takes an optional argument of a custom struct
```
#[wasm_bindgen]
pub fn func(&self, param: Option<&MyStruct>) -> ...
```
requires one to implement `FromWasmAbi` and `OptionFromWasmAbi` for `&MyStruct`, which are not very well documented, and require `unsafe` code and a good understanding of `wasm-bindgen` internals. It looks like something that can be derived automatically.

Alternatively, one would think that the following would work:
```
#[wasm_bindgen]
pub fn func(&self, param: Option<MyStruct>) -> ...
```
This compiles, but leads to unexpected behavior on the JS side (I couldn't find a mention of it in the documentation, which, in fact, states that `Option<T>` parameters are not supported): the first invocation of the binding nullifies the pointer of the JS struct:
```
let param = new MyStruct();
console.log(param);
obj.func(param);
console.log(param);
```
Output:
```
MyStruct {ptr: <some_nonzero_number>}
MyStruct {ptr: 0}
```

**Comment by RReverser** [comment]
For clarity, this issue should probably be rephrased that its specifically for passing `Option` with references. By-value consumption and returns are already implemented and work as expected (as your last example shows).

**Comment by fjarri** [comment]
I must admit I certainly did not expect this behavior on JS side, but I understand how it makes sense if you want to make your JS code behave like Rust. Also, I couldn't find any warning about it in the docs (in fact, `Option<T>` parameters [are explicitly marked as not supported](https://rustwasm.github.io/docs/wasm-bindgen/reference/types/result.html))

**Comment by RReverser** [comment]
@fjarri That table is for `Result<T, JsValue>` specifically. Like for other types in that section, what it means is that `Option<Result<T, JsValue>>` is not supported in params (I admit, it's confusing though, due to usage of `T` in both).

**Comment by RReverser** [comment]
That said, [here](https://rustwasm.github.io/docs/wasm-bindgen/reference/types/exported-rust-types.html) it should certainly be marked as supported in both params and returns.

**Comment by fjarri** [comment]
> Like for other types in that section, what it means is that Option<Result<T, JsValue>> is not supported in params (I admit, it's confusing though, due to usage of T in both).

Ah, I see, my mistake.

**Comment by alexcrichton** [comment]
FWIW `Option<T>` "taking" the value is respecting the Rust-semantics where you're acquiring ownership of a value so it's effectively free'd afterwards, but the only way we have to reflect that in JS is to zero out the pointer. In any case thanks for the issue!

**Comment by fjarri** [comment]
Yes, I figured that. But honestly, I don't see a point in bringing that specific bit of Rust semantic into JS which misses all the other parts that work together with it, like references or the borrow checker.

**Comment by RReverser** [comment]
> But honestly, I don't see a point in bringing that specific bit of Rust semantics into JS which misses all the other parts that work together with it, like references or the borrow checker.

Well it has to do that, otherwise this would be UB on the Rust part. Rust expects that if a parameter is marked to take something by-value, then it won't be accessible again - you're violating that if you still allow access to the same structure on the JS side after the call.

**Comment by fjarri** [comment]
On the other hand, you're violating the JS calling conventions by invalidating the object, so it's rather a question of priorities. I do not mind a *Rust* object inside the bindings not being accessible anymore, but why should it affect the *JS* object?

**Comment by RReverser** [comment]
Because JS object is merely a wrapper for Rust one. It doesn't have a copy of the struct or something like that, it's just a pointer to the Rust side. When Rust frees the data behind such pointer, it would be worse if JS kept allowing access to the "garbage" that now happens to reside there in memory.

**Comment by RReverser** [comment]
If you want JS side to have proper JS objects that are merely copies of the Rust ones, it might be that you want Serde integration instead (e.g. check out https://github.com/cloudflare/serde-wasm-bindgen).

**Comment by fjarri** [comment]
> Because JS object is merely a wrapper for Rust one.

That's where I would disagree. A JS object is a wrapper over a WASM one, that's the reality that a user sees. Rust is merely an implementation detail. There's no call for leaking its semantics into JS, especially those parts of it that don't work well by themselves.

But I agree that this discussion is only tangential to this issue. If `Option<&>` is supported, that will solve the problems that I'm having.

**Comment by fjarri** [comment]
Bump. Any news on this? At the moment the only way to have optional arguments that behave as expected on the JS side (that is, don't get nullified on use) seems to be to use a hacky proc-macro (https://github.com/rustwasm/wasm-bindgen/issues/2231#issuecomment-1169658111). I think it would be a very useful feature to a lot of people to be able to have `Option<&...>` args out of the box.

**Comment by fjarri** [comment]
For anyone who bumps into the same issue: a workaround is possible, see https://docs.rs/wasm-bindgen-derive/latest/wasm_bindgen_derive/#optional-arguments

**Comment by tamaroning** [comment]
Hello, 
There are some hacky solutions provided by https://github.com/rustwasm/wasm-bindgen/issues/2370#issuecomment-1250136858 and https://github.com/rustwasm/wasm-bindgen/issues/2231#issuecomment-1169658111.
However, I hope wasm-bindgen supports `Option<&T>` as a parameter in a safe way.

> FWIW Option<T> "taking" the value is respecting the Rust-semantics where you're acquiring ownership of a value so it's effectively free'd afterwards, but the only way we have to reflect that in JS is to zero out the pointer. In any case thanks for the issue!
https://github.com/rustwasm/wasm-bindgen/issues/2370#issuecomment-735846571

@alexcrichton @RReverser 
I think it is OK not to free arguments if they are of type `Option<&T>`. Could you clarify why it does not work?

**Comment by Enduriel** [comment]
I'm new to the whole wasm-bindgen thing, but I'm trying to use this in a production setting and the fact that I can't use an `Option<&T>` makes designing a clean API significantly more difficult. Therefore, I wanted to add my voice of support to fixing this. The hacky solutions seem to require cloning the underlying struct which shouldn't be necessary.

What is the difficult part of implementing this?

EDIT: Between the lack of TS interoperable enum support, no trait support, no optional references, no easy Rc/Arc support to take the JS approach to memory management, no structs/functions with lifetime parameters, exposing a 'natural' API that converts between rust and JS seems to be very difficult.

**Comment by TomCrypto** [comment]
I too would like to put in a request for this feature. I was surprised that `Option<&T>` didn't just work out of the box, and I couldn't find a rationale in the thread as to why it isn't implemented. The semantics seem clear to me?
